netket/_vmc.py

```diff
@@ -186,6 +186,10 @@ def n_discard(self, n_discard):
             else self._n_samples_node * self._batch_size // 10
         )
 
+    @property
+    def energy_stats(self):
```

Comment:
I don't really see the need for making the energy stats a public property at the moment. (In the internal code, you can use `self._stats` in its place of course.)

> I think that irregardless of this PR it might be nice for users have a way to get only the energy stats from variational.Vmc

I definitely agree. However, `Vmc.run` is primarily a convenience function that runs the VMC optimization loop without the user having to care about the output (which is also why I think including a progress bar makes a lot of sense).
If more control is wanted, the interface for that is `Vmc.iter`. If one does not want to get all observables but just the energy (and, say, the other observables every `n`th step), one could e.g. run
```python
other_obs = [SzSz, SxSx, SySy]
vmc = Vmc(...)
for step in vmc.iter(100):
    print(f"Energy = {vmc.get_observable_stats()}")
    if step % n == 0:
        print(f"Other observables: {vmc.get_observable_stats(other_obs, include_energy=False)}")
```
(For a progress bar in that setting, users of course can add a custom progress bar right now (`for step in tqdm(vmc.iter(100), total=100)`), as I've in fact been doing for a while in some scripts.
Using `add_observable` is only really necessary when one wants to use the simpler `Vmc.run` interface and in that case, having `Vmc.energy_stats` does not add too much value (as the other observables have already been computed anyway, when the users can access that property).

Reply:
Thanks for the detailed comments.

> I don't really see the need for making the energy stats a public property at the moment. (In the internal code, you can use self._stats in its place of course.)

Well since `run` is a method of `variational.Vmc` and `_stats` belongs to `_vmc.Vmc` I was under the impression that by python convention I shouldn't access it directly in `variational.Vmc`.

It was indeed not so clear to me why there are two `Vmc` classes. But as I gather from your comment this is a temporary thing and the goal is to have everything inside `_vmc.Vmc` ? 

> Using add_observable is only really necessary when one wants to use the simpler Vmc.run interface and in that case, having Vmc.energy_stats does not add too much value (as the other observables have already been computed anyway, when the users can access that property).

I'm not sure I follow this comment. As far as I understand, with the current implementation `Vmc.get_observable_stats()` will compute the statistics for all observables which were added  using `add_observables`. What would be the way to access only the energy statistics (assuming that some observables were added) ?

Reply:
> I'm not sure I follow this comment. As far as I understand, with the current implementation Vmc.get_observable_stats() will compute the statistics for all observables which were added using add_observables. 

Indeed, that's true. What I mean is: Since `get_observable_stats(observables=[...])` allows passing in the observables to estimate, it is not necessary to use `add_observable` at all when using the `netket.Vmc.iter` interface (that doesn't have the `run` method altogether), it just provides the default. The only exception is when `Vmc.run` is used, as this of course directly saves the output without giving user code an option to call `get_observable_stats` every step.

The reason for its existence (and for the two separate Vmc classes) is mostly historical: Some time ago, `netket.variational.Vmc` was written in C++. In the process of porting larger parts of NetKet to pure Python, `netket.Vmc` was introduces, which is a pure Python driver. Due to some interface changes (such as introducing a separate `SR` class instead of passing all SR options in the constructor), we could not immediately replace the old driver (as that would be backwards-incompatible), so the current `netket.variational.Vmc` was written (also in Python) to provide the old interface but using the new driver underneath. In the future, it will be deprecated and `netket.Vmc` will be the only driver.

I'd argue that we should also remove `netket.Vmc.add_observable` at that point. (If we keep a `.run`-style method in the new class, it could just take the observables to output as a parameter instead of having those as state of the driver.)

For your PR, that doesn't make too much of a difference (sorry for the long text, but I thought it could help to provide some context). I'd just argue its okay here to use internal interface even in another Python class, as the existence of two classes is just for backwards compatibility and they are closely coupled anyway.